src/components/ThemeProvider/ThemeProvider.tsx

```diff
@@ -1,109 +1,45 @@
-import React, {useMemo, useEffect, useContext} from 'react';
-import DefaultThemeColors from '@shopify/polaris-tokens/dist-modern/theme/base.json';
+import React from 'react';
 
-import {
-  Theme,
-  ThemeContext,
-  ThemeConfig,
-  buildThemeContext,
-  buildCustomProperties,
-  toString,
-  Tokens,
-} from '../../utilities/theme';
+import type {ColorScheme} from './tokens';
+import {customProperties} from './customProperties';
+
+const styleSheet = document.createElement('style');
+styleSheet.textContent = customProperties;
+styleSheet.dataset.polarisCustomProperties = '👋';
+
+// This variable is intentionally in module scope to ensure
+// Polaris custom properties are only inject one time.
+let injectedCustomProperties = false;
 
 export interface ThemeProviderProps {
-  /** Custom logos and colors provided to select components */
-  theme?: ThemeConfig;
-  /**
-   * By default, Polaris avoids re-declaring custom properties within the same React tree
-   * This prop ensures that the CSS custom properties are always rendered. This is useful
-   * for components such as portals that render outside of the root DOM node
-   */
-  alwaysRenderCustomProperties?: boolean;
-  /** The content to display */
+  /** Custom logos and colors provided to select components. */
+  colorScheme?: ColorScheme;
+  /** The content to display. */
   children?: React.ReactNode;
+  /** Class name applied to the root element. */
+  className?: string;
+  /** Element used for the root node. */
+  as?: React.ElementType;
 }
 
-export function ThemeProvider({
-  theme: themeConfig = {},
-  alwaysRenderCustomProperties = false,
-  children,
-}: ThemeProviderProps) {
-  const parentContext = useContext(ThemeContext);
-  const isParentThemeProvider = parentContext === undefined;
-
-  const parentColorScheme =
-    parentContext && parentContext.colorScheme && parentContext.colorScheme;
-  const parentColors =
-    parentContext && parentContext.colors && parentContext.colors;
-
-  const [customProperties, theme] = useMemo(() => {
-    const {colors, colorScheme, ...rest} = themeConfig;
-
-    const processedThemeConfig = {
-      ...rest,
-      ...{colorScheme: getColorScheme(colorScheme, parentColorScheme)},
-      colors: {
-        ...(isParentThemeProvider && DefaultThemeColors),
-        ...(parentColors != null && parentColors),
-        ...colors,
-      },
-    };
-
-    const customProperties = buildCustomProperties(
-      processedThemeConfig,
-      Tokens,
-    );
+export function ThemeProvider(props: ThemeProviderProps) {
+  const {
+    as: Component = 'div',
+    children,
+    className,
+    colorScheme = 'light',
+  } = props;
 
-    const theme = buildThemeContext(processedThemeConfig, customProperties);
+  React.useEffect(() => {
+    if (injectedCustomProperties) return undefined;
+    injectedCustomProperties = true;
```

Comment:
Alternatively, we could determine if Polaris custom properties have already been injected with a `data-*` attribute on the style tag (instead of the module scoped `injectedCustomProperties` variable). Such as:
```jsx
const style = document.createElement('style');
style.dataset.polarisCustomProperties = 'false';

...

React.useEffect(() => {
  if (JSON.parse(style.dataset.polarisCustomProperties)) return undefined;
  style.dataset.polarisCustomProperties = 'true';
  document.head.appendChild(style);
}, []);
```

Reply:
Discussed offline, the module scoped `injectCustomProperties` variable is sufficient.